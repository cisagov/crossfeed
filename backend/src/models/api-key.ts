import {
  Entity,
  Column,
  PrimaryGeneratedColumn,
  UpdateDateColumn,
  CreateDateColumn,
  BaseEntity,
  ManyToOne
} from 'typeorm';
import { User } from './';

@Entity()
export class ApiKey extends BaseEntity {
  @PrimaryGeneratedColumn('uuid')
  id: string;

  @CreateDateColumn()
  createdAt: Date;

  @UpdateDateColumn()
  updatedAt: Date;

  @ManyToOne((type) => User, (user) => user.apiKeys, {
    onDelete: 'CASCADE',
    onUpdate: 'CASCADE'
  })
  user: User;

  @Column({
    type: 'timestamp',
    nullable: true
  })
  lastUsed: Date | null;

  @Column({
    // length based on hex formatted sha256
    length: 64,
    type: 'varchar'
  })
  hashedKey: string;

  @Column({
    length: 4,
    type: 'varchar'
  })
  lastFour: string;
}
