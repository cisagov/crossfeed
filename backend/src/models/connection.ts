import { createConnection, Connection } from 'typeorm';
import {
  Domain,
  Service,
  Vulnerability,
  Scan,
  Organization,
  User,
  Role,
  ScanTask
} from '.';

let connection: Connection | null = null;

const connectDb = async (logging?: boolean) => {
  const connection = createConnection({
    type: 'postgres',
    host: process.env.DB_HOST,
    port: parseInt(process.env.DB_PORT ?? ''),
    username: process.env.DB_USERNAME,
    password: process.env.DB_PASSWORD,
    database: process.env.DB_NAME,
    entities: [
      Domain,
      Service,
      Vulnerability,
      Scan,
      Organization,
      User,
      Role,
      ScanTask
    ],
    synchronize: false,
    name: 'default',
    dropSchema: false,
    logging: logging ?? false
  });
  return connection;
};

export const connectToDatabase = async (logging?: boolean) => {
  if (!connection?.isConnected) {
    connection = await connectDb(logging);
  }
  if (typeof jest === 'undefined') {
    console.log('=> DB Connected');
  }
  return connection;
};
