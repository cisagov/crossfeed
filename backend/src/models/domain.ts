import {
  Entity,
  Index,
  Column,
  OneToMany,
  PrimaryGeneratedColumn,
  UpdateDateColumn,
  CreateDateColumn,
  BaseEntity,
  BeforeInsert,
  ManyToOne
} from 'typeorm';
import { Service } from './service';
import { Organization } from './organization';
import { Vulnerability } from './vulnerability';
import { Scan } from './scan';
import { Webpage } from './webpage';

@Entity()
@Index(['name', 'organization'], { unique: true })
export class Domain extends BaseEntity {
  @PrimaryGeneratedColumn('uuid')
  id: string;

  @CreateDateColumn()
  createdAt: Date;

  @UpdateDateColumn()
  updatedAt: Date;

  /** When this model was last synced with Elasticsearch. */
  @Column({
    type: 'timestamp',
    nullable: true
  })
  syncedAt: Date | null;

  @Column({
    nullable: true
  })
  ip: string;

  /** Associated root domain that led to the discovery of this domain. */
  @Column({
    nullable: true
  })
  fromRootDomain: string;

  @ManyToOne((type) => Scan, {
    onDelete: 'SET NULL',
    onUpdate: 'CASCADE'
  })
  discoveredBy: Scan;

  @Column({
    length: 512
  })
  reverseName: string;

  @Column({
    length: 512
  })
  name: string;

  @OneToMany((type) => Service, (service) => service.domain)
  services: Service[];

  @OneToMany((type) => Vulnerability, (vulnerability) => vulnerability.domain)
  vulnerabilities: Vulnerability[];

  @OneToMany((type) => Webpage, (webpage) => webpage.domain)
  webpages: Service[];

  @ManyToOne((type) => Organization, { onDelete: 'CASCADE', nullable: false })
  organization: Organization;

  @Column({
    length: 512,
    nullable: true,
    type: 'varchar'
  })
  screenshot: string | null;

  @Column({
    nullable: true,
    type: 'varchar'
  })
  country: string | null;

  @Column({
    nullable: true,
    type: 'varchar'
  })
  asn: string | null;

  @Column({
    default: false
  })
  cloudHosted: boolean;

  /** SSL Certificate information  */
  @Column({
    type: 'jsonb',
    nullable: true
  })
  ssl: {
    issuerOrg: string;
    issuerCN: string;
    validFrom: string;
    validTo: string;
    protocol: string;
    altNames: string;
    bits: string;
    fingerprint: string;
  } | null;

  /** Censys Certificates results */
  @Column({
    type: 'jsonb',
    default: {}
  })
  censysCertificatesResults: {
    [x: string]: any;
  };

  @BeforeInsert()
  setLowerCase() {
    this.name = this.name.toLowerCase();
  }

  @BeforeInsert()
  setReverseName() {
    this.reverseName = this.name.split('.').reverse().join('.');
  }
}
