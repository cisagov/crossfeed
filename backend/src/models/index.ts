export * from './domain';
export * from './service';
export * from './connection';
export * from './vulnerability';
export * from './scan';
export * from './organization';
export * from './user';
export * from './role';
export * from './scan-task';
