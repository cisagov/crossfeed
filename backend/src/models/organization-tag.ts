import {
  Entity,
  Index,
  PrimaryGeneratedColumn,
  UpdateDateColumn,
  CreateDateColumn,
  BaseEntity,
  ManyToMany,
  JoinTable,
  Column
} from 'typeorm';
import { Organization, Scan } from '.';

@Entity()
@Index(['name'], { unique: true })
export class OrganizationTag extends BaseEntity {
  @PrimaryGeneratedColumn('uuid')
  id: string;

  @CreateDateColumn()
  createdAt: Date;

  @UpdateDateColumn()
  updatedAt: Date;

  @Column()
  name: string;

  /**
   * Organizations that are labeled with this tag
   */
  @ManyToMany((type) => Organization, (organization) => organization.tags, {
    onDelete: 'CASCADE',
    onUpdate: 'CASCADE'
  })
  @JoinTable()
  organizations: Organization[];

  /**
   * Scans that have this tag enabled, and will run against all tagged organizations
   */
  @ManyToMany((type) => Scan, (scan) => scan.tags, {
    onDelete: 'CASCADE',
    onUpdate: 'CASCADE'
  })
  scans: Scan[];
}
