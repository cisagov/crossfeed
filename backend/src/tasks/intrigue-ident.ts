import { Domain, Service } from '../models';
import { CommandOptions } from './ecs-client';
import getLiveWebsites from './helpers/getLiveWebsites';
import PQueue from 'p-queue';
import * as buffer from 'buffer';
import { spawnSync } from 'child_process';

const intrigueIdent = async (domain: Domain): Promise<void> => {
  console.log('Domain', domain.name);
  const ports = domain.services.map((service) => service.port);
  let service: Service;
  if (ports.includes(443))
    service = domain.services.find((service) => service.port === 443)!;
  else service = domain.services.find((service) => service.port === 80)!;
  const url =
    service.port === 443 ? `https://${domain.name}` : `http://${domain.name}`;
  const { stdout, stderr, status } = spawnSync(
    'intrigue-ident',
    ['--uri', url, '--json'],
    {
      env: {
        ...process.env,
        HTTP_PROXY: process.env.GLOBAL_AGENT_HTTP_PROXY,
        HTTPS_PROXY: process.env.GLOBAL_AGENT_HTTP_PROXY
      }
    }
  );
  if (stderr.toString()) {
    console.error('stderr', stderr.toString());
  }
  if (status !== 0) {
    console.error('IntrigueIdent failed');
    return;
  }
  const output = stdout.toString();
  const { fingerprint, content } = JSON.parse(
    output.substring(output.indexOf('{'))
  );
  service.intrigueIdentResults = { fingerprint, content };
  await service.save();
};

export const handler = async (commandOptions: CommandOptions) => {
  const { organizationId, organizationName } = commandOptions;

  console.log('Running intrigueIdent on organization', organizationName);

  const liveWebsites = await getLiveWebsites(organizationId!);
  const queue = new PQueue({ concurrency: 5 });
  await Promise.all(
    liveWebsites.map((site) => queue.add(() => intrigueIdent(site)))
  );

  console.log(`IntrigueIdent finished for ${liveWebsites.length} domains`);
};
