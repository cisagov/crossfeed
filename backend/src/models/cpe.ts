import {
  Entity,
  PrimaryGeneratedColumn,
  Column,
  ManyToMany,
  BaseEntity,
  Unique
} from 'typeorm';
import { Cve } from './cve';

@Entity()
@Unique(['name', 'version', 'vendor'])
export class Cpe extends BaseEntity {
  @PrimaryGeneratedColumn('uuid')
  id: string;

  @Column()
  name: string;

  @Column()
  version: string;

  @Column()
  vendor: string;

  @Column()
  lastSeenAt: Date;

  @ManyToMany(() => Cve, (cve) => cve.cpes)
  cves: Cve[];
}
