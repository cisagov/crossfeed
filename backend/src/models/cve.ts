import {
  Entity,
  PrimaryGeneratedColumn,
  Column,
  ManyToMany,
  BaseEntity,
  JoinTable,
  Unique
} from 'typeorm';
import { Cpe } from './cpe';

//TODO: Refactor column names to camelCase to match the rest of the codebase?
@Entity()
@Unique(['name'])
export class Cve extends BaseEntity {
  @PrimaryGeneratedColumn('uuid')
  id: string;

  @Column({ nullable: true })
  name: string;

  @Column({ nullable: true })
  publishedAt: Date;

  @Column({ nullable: true })
  modifiedAt: Date;

  @Column({ nullable: true })
  status: string;

  @Column({ nullable: true })
  description: string;

  @Column({ nullable: true })
  cvssV2Source: string;

  @Column({ nullable: true })
  cvssV2Type: string;

  @Column({ nullable: true })
  cvssV2Version: string;

  @Column({ nullable: true })
  cvssV2VectorString: string;

  @Column({ nullable: true })
  cvssV2BaseScore: string;

  @Column({ nullable: true })
  cvssV2BaseSeverity: string;

  @Column({ nullable: true })
  cvssV2ExploitabilityScore: string;

  @Column({ nullable: true })
  cvssV2ImpactScore: string;

  @Column({ nullable: true })
  cvssV3Source: string;

  @Column({ nullable: true })
  cvssV3Type: string;

  @Column({ nullable: true })
  cvssV3Version: string;

  @Column({ nullable: true })
  cvssV3VectorString: string;

  @Column({ nullable: true })
  cvssV3BaseScore: string;

  @Column({ nullable: true })
  cvssV3BaseSeverity: string;

  @Column({ nullable: true })
  cvssV3ExploitabilityScore: string;

  @Column({ nullable: true })
  cvssV3ImpactScore: string;

  @Column({ nullable: true })
  cvssV4Source: string;

  @Column({ nullable: true })
  cvssV4Type: string;

  @Column({ nullable: true })
  cvssV4Version: string;

  @Column({ nullable: true })
  cvssV4VectorString: string;

  @Column({ nullable: true })
  cvssV4BaseScore: string;

  @Column({ nullable: true })
  cvssV4BaseSeverity: string;

  @Column({ nullable: true })
  cvssV4ExploitabilityScore: string;

  @Column({ nullable: true })
  cvssV4ImpactScore: string;

  @Column('simple-array', { nullable: true })
  weaknesses: string[];

  @Column('simple-array', { nullable: true })
  references: string[];

  @ManyToMany(() => Cpe, (cpe) => cpe.cves, {
    cascade: true
  })
  @JoinTable()
  cpes: Cpe[];
}
