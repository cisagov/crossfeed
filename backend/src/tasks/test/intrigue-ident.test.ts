import { handler as intrigueIdent } from '../intrigue-ident';
import { connectToDatabase, Organization, Service, Domain } from '../../models';
import { spawnSync } from 'child_process';

jest.mock('child_process', () => ({
  spawnSync: jest.fn()
}));

const mockIntrigueIdentResponse = (response) => {
  (spawnSync as jest.Mock).mockImplementation(() => ({
    status: 0,
    stderr: '',
    stdout:
      `
Fingerprint:
- Bootstrap Bootstrap   - boostrap css (CPE: cpe:2.3:a:bootstrap:bootstrap::) (Tags: ["Web Framework"]) (Hide: false)
- Apache HTTP Server   - Apache server header w/o version (CPE: cpe:2.3:a:apache:http_server::) (Tags: ["Web Server"]) (Hide: false)
- Apache HTTP Server   - Apache web server - server header - no version (CPE: cpe:2.3:a:apache:http_server::) (Tags: ["Web Server"]) (Hide: false)
- Drupal Drupal 8  - Drupal headers (CPE: cpe:2.3:a:drupal:drupal:8:) (Tags: ["CMS"]) (Hide: false)
- NewRelic NewRelic   - NewRelic tracking code (CPE: cpe:2.3:s:newrelic:newrelic::) (Tags: ["APM", "Javascript"]) (Hide: false)
` + JSON.stringify(response, null, 2)
  }));
};

describe('intrigue ident', () => {
  beforeAll(async () => {
    await connectToDatabase();
  });
  let organization;
  beforeEach(async () => {
    organization = await Organization.create({
      name: 'test-' + Math.random(),
      rootDomains: ['test-' + Math.random()],
      ipBlocks: [],
      isPassive: false
    }).save();
  });
  test('basic test', async () => {
    mockIntrigueIdentResponse({
      url: 'https://www.cisa.gov/',
      fingerprint: [
        {
          type: 'fingerprint',
          vendor: 'Bootstrap',
          product: 'Bootstrap',
          version: '',
          update: '',
          tags: ['Web Framework'],
          match_type: 'content_body',
          match_details: 'boostrap css',
          hide: false,
          cpe: 'cpe:2.3:a:bootstrap:bootstrap::',
          issue: null,
          task: null,
          inference: false
        },
        {
          type: 'fingerprint',
          vendor: 'Apache',
          product: 'HTTP Server',
          version: '',
          update: '',
          tags: ['Web Server'],
          match_type: 'content_headers',
          match_details: 'Apache server header w/o version',
          hide: false,
          cpe: 'cpe:2.3:a:apache:http_server::',
          issue: null,
          task: null,
          inference: false
        },
        {
          type: 'fingerprint',
          vendor: 'Apache',
          product: 'HTTP Server',
          version: '',
          update: '',
          tags: ['Web Server'],
          match_type: 'content_headers',
          match_details: 'Apache web server - server header - no version',
          hide: false,
          cpe: 'cpe:2.3:a:apache:http_server::',
          issue: null,
          task: null,
          inference: false
        },
        {
          type: 'fingerprint',
          vendor: 'Drupal',
          product: 'Drupal',
          version: '8',
          update: '',
          tags: ['CMS'],
          match_type: 'content_headers',
          match_details: 'Drupal headers',
          hide: false,
          cpe: 'cpe:2.3:a:drupal:drupal:8:',
          issue: null,
          task: null,
          inference: false
        },
        {
          type: 'fingerprint',
          vendor: 'NewRelic',
          product: 'NewRelic',
          version: '',
          update: '',
          tags: ['APM', 'Javascript'],
          match_type: 'content_body',
          match_details: 'NewRelic tracking code',
          hide: false,
          cpe: 'cpe:2.3:s:newrelic:newrelic::',
          issue: null,
          task: null,
          inference: null
        }
      ],
      content: [
        {
          type: 'content',
          name: 'Location Header',
          hide: null,
          issue: null,
          task: null,
          result: null
        },
        {
          type: 'content',
          name: 'Directory Listing Detected',
          hide: null,
          issue: null,
          task: null,
          result: false
        },
        {
          type: 'content',
          name: 'Form Detected',
          hide: false,
          issue: false,
          task: null,
          result: true
        },
        {
          type: 'content',
          name: 'File Upload Form Detected',
          hide: null,
          issue: null,
          task: null,
          result: false
        },
        {
          type: 'content',
          name: 'Email Addresses Detected',
          hide: false,
          issue: false,
          task: null,
          result: []
        },
        {
          type: 'content',
          name: 'Authentication - HTTP',
          hide: null,
          issue: null,
          task: null,
          result: false
        },
        {
          type: 'content',
          name: 'Authentication - NTLM',
          hide: null,
          issue: null,
          task: null,
          result: false
        },
        {
          type: 'content',
          name: 'Authentication - Forms',
          hide: null,
          issue: null,
          task: null,
          result: false
        },
        {
          type: 'content',
          name: 'Authentication - Session Identifier',
          hide: null,
          issue: null,
          task: null,
          result: false
        },
        {
          type: 'content',
          name: 'Access-Control-Allow-Origin Header',
          hide: null,
          issue: null,
          task: null,
          result: false
        },
        {
          type: 'content',
          name: 'P3P Header',
          hide: null,
          issue: null,
          task: null,
          result: false
        },
        {
          type: 'content',
          name: 'X-Frame-Options Header',
          hide: false,
          issue: false,
          task: null,
          result: true
        },
        {
          type: 'content',
          name: 'X-XSS-Protection Header',
          hide: null,
          issue: null,
          task: null,
          result: false
        },
        {
          type: 'content',
          name: 'Google Analytics',
          hide: null,
          issue: null,
          task: null,
          result: null
        },
        {
          type: 'content',
          name: 'Google Site Verification',
          hide: null,
          issue: null,
          task: null,
          result: null
        },
        {
          type: 'content',
          name: 'MyWebStats',
          hide: null,
          issue: null,
          task: null,
          result: null
        }
      ],
      responses: [],
      initial_checks: [
        {
          url: 'https://www.cisa.gov/',
          count: 676
        }
      ],
      followon_checks: []
    });
    const domain = await Domain.create({
      organization,
      name: 'www.cisa.gov',
      ip: '0.0.0.0'
    }).save();
    let service = await Service.create({
      service: 'https',
      port: 443,
      domain
    }).save();
    await intrigueIdent({
      organizationId: organization.id,
      organizationName: 'organizationName',
      scanId: 'scanId',
      scanName: 'scanName',
      scanTaskId: 'scanTaskId',
      chunkNumber: 0,
      numChunks: 1
    });
    service = await Service.findOneOrFail(service.id);
    expect(service.intrigueIdentResults).toMatchSnapshot();
    expect(service.products).toMatchSnapshot();
  });
});
