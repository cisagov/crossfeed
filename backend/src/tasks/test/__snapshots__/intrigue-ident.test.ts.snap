// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`intrigue ident basic test 1`] = `
Object {
  "content": Array [
    Object {
      "hide": null,
      "issue": null,
      "name": "Location Header",
      "result": null,
      "task": null,
      "type": "content",
    },
    Object {
      "hide": null,
      "issue": null,
      "name": "Directory Listing Detected",
      "result": false,
      "task": null,
      "type": "content",
    },
    Object {
      "hide": false,
      "issue": false,
      "name": "Form Detected",
      "result": true,
      "task": null,
      "type": "content",
    },
    Object {
      "hide": null,
      "issue": null,
      "name": "File Upload Form Detected",
      "result": false,
      "task": null,
      "type": "content",
    },
    Object {
      "hide": false,
      "issue": false,
      "name": "Email Addresses Detected",
      "result": Array [],
      "task": null,
      "type": "content",
    },
    Object {
      "hide": null,
      "issue": null,
      "name": "Authentication - HTTP",
      "result": false,
      "task": null,
      "type": "content",
    },
    Object {
      "hide": null,
      "issue": null,
      "name": "Authentication - NTLM",
      "result": false,
      "task": null,
      "type": "content",
    },
    Object {
      "hide": null,
      "issue": null,
      "name": "Authentication - Forms",
      "result": false,
      "task": null,
      "type": "content",
    },
    Object {
      "hide": null,
      "issue": null,
      "name": "Authentication - Session Identifier",
      "result": false,
      "task": null,
      "type": "content",
    },
    Object {
      "hide": null,
      "issue": null,
      "name": "Access-Control-Allow-Origin Header",
      "result": false,
      "task": null,
      "type": "content",
    },
    Object {
      "hide": null,
      "issue": null,
      "name": "P3P Header",
      "result": false,
      "task": null,
      "type": "content",
    },
    Object {
      "hide": false,
      "issue": false,
      "name": "X-Frame-Options Header",
      "result": true,
      "task": null,
      "type": "content",
    },
    Object {
      "hide": null,
      "issue": null,
      "name": "X-XSS-Protection Header",
      "result": false,
      "task": null,
      "type": "content",
    },
    Object {
      "hide": null,
      "issue": null,
      "name": "Google Analytics",
      "result": null,
      "task": null,
      "type": "content",
    },
    Object {
      "hide": null,
      "issue": null,
      "name": "Google Site Verification",
      "result": null,
      "task": null,
      "type": "content",
    },
    Object {
      "hide": null,
      "issue": null,
      "name": "MyWebStats",
      "result": null,
      "task": null,
      "type": "content",
    },
  ],
  "fingerprint": Array [
    Object {
      "cpe": "cpe:2.3:a:bootstrap:bootstrap::",
      "hide": false,
      "inference": false,
      "issue": null,
      "match_details": "boostrap css",
      "match_type": "content_body",
      "product": "Bootstrap",
      "tags": Array [
        "Web Framework",
      ],
      "task": null,
      "type": "fingerprint",
      "update": "",
      "vendor": "Bootstrap",
      "version": "",
    },
    Object {
      "cpe": "cpe:2.3:a:apache:http_server::",
      "hide": false,
      "inference": false,
      "issue": null,
      "match_details": "Apache server header w/o version",
      "match_type": "content_headers",
      "product": "HTTP Server",
      "tags": Array [
        "Web Server",
      ],
      "task": null,
      "type": "fingerprint",
      "update": "",
      "vendor": "Apache",
      "version": "",
    },
    Object {
      "cpe": "cpe:2.3:a:apache:http_server::",
      "hide": false,
      "inference": false,
      "issue": null,
      "match_details": "Apache web server - server header - no version",
      "match_type": "content_headers",
      "product": "HTTP Server",
      "tags": Array [
        "Web Server",
      ],
      "task": null,
      "type": "fingerprint",
      "update": "",
      "vendor": "Apache",
      "version": "",
    },
    Object {
      "cpe": "cpe:2.3:a:drupal:drupal:8:",
      "hide": false,
      "inference": false,
      "issue": null,
      "match_details": "Drupal headers",
      "match_type": "content_headers",
      "product": "Drupal",
      "tags": Array [
        "CMS",
      ],
      "task": null,
      "type": "fingerprint",
      "update": "",
      "vendor": "Drupal",
      "version": "8",
    },
    Object {
      "cpe": "cpe:2.3:s:newrelic:newrelic::",
      "hide": false,
      "inference": null,
      "issue": null,
      "match_details": "NewRelic tracking code",
      "match_type": "content_body",
      "product": "NewRelic",
      "tags": Array [
        "APM",
        "Javascript",
      ],
      "task": null,
      "type": "fingerprint",
      "update": "",
      "vendor": "NewRelic",
      "version": "",
    },
  ],
}
`;

exports[`intrigue ident basic test 2`] = `
Array [
  Object {
    "cpe": "cpe:/a:bootstrap:bootstrap::",
    "name": "Bootstrap",
    "tags": Array [
      "Web Framework",
    ],
    "vendor": "Bootstrap",
  },
  Object {
    "cpe": "cpe:/a:apache:http_server::",
    "name": "HTTP Server",
    "tags": Array [
      "Web Server",
    ],
    "vendor": "Apache",
  },
  Object {
    "cpe": "cpe:/a:drupal:drupal:8:",
    "name": "Drupal",
    "tags": Array [
      "CMS",
    ],
    "vendor": "Drupal",
    "version": "8",
  },
  Object {
    "cpe": "cpe:/s:newrelic:newrelic::",
    "name": "NewRelic",
    "tags": Array [
      "APM",
      "Javascript",
    ],
    "vendor": "NewRelic",
  },
]
`;
